staging/src/k8s.io/apimachinery/pkg/runtime/codec.go

```diff
@@ -45,14 +46,23 @@ func NewCodec(e Encoder, d Decoder) Codec {
 
 // Encode is a convenience wrapper for encoding to a []byte from an Encoder
 func Encode(e Encoder, obj Object) ([]byte, error) {
-	// TODO: reuse buffer
-	buf := &bytes.Buffer{}
+	buf := bufPool.Get().(*bytes.Buffer)
+	defer bufPool.Put(buf)
+
+	buf.Reset()
 	if err := e.Encode(obj, buf); err != nil {
 		return nil, err
 	}
 	return buf.Bytes(), nil
```

Comment:
Do you have any suggest or can I  return `append([]byte(nil), buf.Bytes()...), nil` instead of `buf.Bytes(), nil` @odinuge

Reply:
If the bufPool is `sync.Pool`, get buffer from sync.Pool is safe by multiple goroutine simultaneously(https://pkg.go.dev/sync#Pool), but I think if write a object to the buffer could cause race data condition, because the encode function is not safe.

Reply:
Its been a while since I poked at the byte buffers, but to properly ensure that the returned value is never changed, you can copy it into a new slice. This will however cause _more_ allocations that the current code does when executed once, and since I'm not familiar with how the code is used, I'm not sure what the best or correct solution is here; so defer that to some others..

```golang
func Encode(e Encoder, obj Object) ([]byte, error) {
	buf := bufPool.Get().(*bytes.Buffer)
	defer bufPool.Put(buf)
	buf.Reset()

	if err := e.Encode(obj, buf); err != nil {
		return nil, err
	}
	ret := make([]byte, buf.Len())
	copy(ret, buf.Bytes())
	return ret, nil
}
```

Reply:
thank you , will wait and see what others think.

Reply:
you mean ： https://github.com/golang/go/issues/13671

Reply:
I think 
`func Encode(e Encoder, obj Object) error {
	buf, ok := bufPool.Get().(*bytes.Buffer)
	if !ok {
		return errors.New("bufPool not *bytes.Buffer")
	}
	if buf == nil {
		return errors.New("buf is nil")
	}
	if err := e.Encode(obj, buf); err != nil {
		return err
	}
	if buf.Len() < 1024 {
		buf.Reset()
		bufPool.Put(buf)
		return nil
	}
	return nil
}`

Reply:
I think 
``` go
func Encode(e Encoder, obj Object) error {
	buf, ok := bufPool.Get().(*bytes.Buffer)
	if !ok {
		return errors.New("bufPool not *bytes.Buffer")
	}
	if buf == nil {
		return errors.New("buf is nil")
	}
	if err := e.Encode(obj, buf); err != nil {
		return err
	}
	if buf.Len() < 1024 {
		buf.Reset()
		bufPool.Put(buf)
		return nil
	}
	return nil
}
```

Reply:
I think 
```
func Encode(e Encoder, obj Object) ([]byte, error) {
	buf, ok := bufPool.Get().(*bytes.Buffer)
	if !ok {
		return nil, errors.New("bufPool not *bytes.Buffer")
	}
	if buf == nil {
		return nil, errors.New("buf is nil")
	}
	if err := e.Encode(obj, buf); err != nil {
		return nil, err
	}
	ret := make([]byte, buf.Len())
	copy(ret, buf.Bytes())
	if buf.Len() < 1024 {
		buf.Reset()
		bufPool.Put(buf)
		return ret, nil
	}
	return ret, nil
}

```

Reply:
there's a data race risk here
```go
func main() {
	pool := sync.Pool{
		New: func() interface{} {
			return new(bytes.Buffer)
		},
	}

	buf :=pool.Get().(*bytes.Buffer)
	buf.WriteString("hello")

	arrayRef := buf.Bytes()

	pool.Put(buf)

	buf = pool.Get().(*bytes.Buffer)
	buf.Reset()

	buf.WriteString("world")

	pool.Put(buf)

	fmt.Println(string(arrayRef))
	// Output: world
}
```

Reply:
according to godoc:

> Bytes returns a slice of length b.Len() holding the unread portion of the buffer. The slice is valid for use only until the next buffer modification (that is, only until the next call to a method like Read, Write, Reset, or Truncate)

My read of the godoc is that the bytes returned from the buffer are unsafe to use without copying, which largely negates the point of reusing the buffer.

I think we should remove the TODO rather than try to reuse the buffer here.

Reply:
ok , just remove the todo